#4017: Editing for Kafka ResetOffset example (#4019)
knative/docs: docs/eventing/samples/kafka/resetoffset/README.md

# ResetOffset Example

Kafka-backed Channels differ from other Channel implementations because they provide temporal persistence of events.

Event persistence enables certain advanced use cases to be supported, such as the ability to replay prior events by repositioning their current location in the event stream. This feature is useful when attempting to recover from unexpected Subscriber downtime.

> The [ResetOffset custom resource definition (CRD)](https://github.com/knative-sandbox/eventing-kafka/tree/main/config/command/resetoffset) exposes the ability to manipulate the location of the ConsumerGroup Offsets in the event stream of a given Knative Subscription. Without the ResetOffset CRD, you must manually stop ConsumerGroups and manipulating Offsets.

Reviewer: ```suggestion
The [ResetOffset custom resource definition (CRD)](https://github.com/knative-sandbox/eventing-kafka/tree/main/config/command/resetoffset) exposes the ability to manipulate the location of the ConsumerGroup Offsets in the event stream of a given Knative Subscription. Without the ResetOffset CRD, you must manually stop ConsumerGroups and manipulate Offsets.
```

!!! note
    ResetOffsets are currently only supported by the Distributed KafkaChannel
    implementation.

!!! warning

>     Repositioning the ConsumerGroup Offsets will impact the event ordering and

Reviewer: ```suggestion
    Repositioning the ConsumerGroup Offsets impacts the event ordering and
```

    is intended for failure recovery scenarios. This capability needs to be used
    with caution only after reviewing the [CRD documentation](https://github.com/knative-sandbox/eventing-kafka/tree/main/config/command/resetoffset).

## Prerequisites

- A Kubernetes cluster with the [Kafka Channel implementation](https://knative.dev/docs/eventing/channels/channels-crds/) installed.
- A valid KafkaChannel resource and Subscription exist.

!!! note

>     The ResetOffset CRD is a single-use operation, and should be

Reviewer: ```suggestion
    The ResetOffset CRD is a single-use operation, and must be
```

Author: I actually asked the same thing, but this is a recommendation, not mandatory AFAIK, so I'll leave it as "should" for the moment, even though I know we'd prefer to not do that...

    deleted from the cluster once it has been executed.

## Repositioning offsets to the oldest available event

The following ResetOffset instance moves the offsets back to the oldest
available event in the Kafka Topic retention window:

```yaml
apiVersion: kafka.eventing.knative.dev/v1alpha1
kind: ResetOffset
metadata:
  name: <resetoffset-name>
  namespace: <namespace>
spec:
  offset:
    time: earliest
  ref:
    apiVersion: messaging.knative.dev/v1
    kind: Subscription

>     namespace: <subcription-namespace>

Reviewer: ```suggestion
    namespace: <subscription-namespace>
```

    name: <subscription-name>
```

Where:

- `<resetoffset-name>` is the name of the ResetOffset CRD.
- `<namespace>` is the namespace where you want to create the ResetOffset CRD.

> - `<subcription-namespace>` is the namespace where your Subscription exists.

Reviewer: ```suggestion
- `<subscription-namespace>` is the namespace where your Subscription exists.
```

- `<subscription-name>` is the name of the Subscription.

## Repositioning offsets to a specific point in the event stream

The following ResetOffset instance moves the offsets back to the specified
point in the Kafka Topic retention window.

```yaml
apiVersion: kafka.eventing.knative.dev/v1alpha1
kind: ResetOffset
metadata:
  name: <resetoffset-name>
  namespace: <namespace>
spec:
  offset:
    time: <offset-time>
  ref:
    apiVersion: messaging.knative.dev/v1
    kind: Subscription

>     namespace: <subcription-namespace>

Reviewer: ```suggestion
    namespace: <subscription-namespace>
```

    name: <subscription-name>
```

Where:

- `<resetoffset-name>` is the name of the ResetOffset CRD.
- `<offset-time>` is the specified offset time. You will need to set an [RFC3339](https://datatracker.ietf.org/doc/html/rfc3339) timestamp relative to your use case, for example `"2021-06-17T17:30:00Z"`.
- `<namespace>` is the namespace where you want to create the ResetOffset CRD.

> - `<subcription-namespace>` is the namespace where your Subscription exists.

Reviewer: ```suggestion
- `<subscription-namespace>` is the namespace where your Subscription exists.
```

Offsets.
# ResetOffset Example

!!! note
    Repositioning the ConsumerGroup Offsets will impact the event ordering and
    is intended for failure recovery scenarios. This capability needs to be used
    with caution only after reviewing the CRD documentation linked above.
Kafka-backed Channels differ from other Channel implementations because they provide temporal persistence of events.

Event persistence enables certain advanced use cases to be supported, such as the ability to replay prior events by repositioning their current location in the event stream. This feature is useful when attempting to recover from unexpected Subscriber downtime.

The [ResetOffset custom resource definition (CRD)](https://github.com/knative-sandbox/eventing-kafka/tree/main/config/command/resetoffset) exposes the ability to manipulate the location of the ConsumerGroup Offsets in the event stream of a given Knative Subscription. Without the ResetOffset CRD, you must manually stop ConsumerGroups and manipulating Offsets.

!!! note
    ResetOffsets are currently only supported by the Distributed KafkaChannel
    implementation. It is expected that the Consolidated KafkaChannel

>     implementation will support them in the near future.

Author: @travis-minke-sap we don't really talk about what might happen in the future in docs, just FYI 🙂

Reviewer 2: 👍

>     <!--TODO: Explain why?-->

Author: @travis-minke-sap do you know why this is? I haven't heard of any other single use CRDs in Knative

Reviewer 2: Yes, this resource is a bit different.  The feature went through pretty extensive investigation / design / review in order to achieve agreement that this was the "best" way forward.  Essentially we are specifying a one-time operation (e.g. move the offsets once right now and then be done) in the standard Kubernetes declarative fashion.  The README linked at the top of this examples page provides more information.

Maybe something like the following would be better?

```
The ResetOffset CRD is a single-use operation that should be monitored and removed once it has successfully completed.  Previously successful ResetOffset instances will not be re-executed, but are clutter that must be ignored.
```

Author: I think it's actually less confusing as is 🙂 I was just curious as to why
